### ISSUE TITLE
visitor-as: Implement AS transform

### ISSUE THREAD

**Issue by willemneal** [near/lang-DAO-grants#7]
Visitor-as is an important library necessary for functionality of near-sdk-as since it allows macros-like code injections. Currently it is mostly written in TypeScript, since AssemblyScript compiler front-end is still mostly written in JavaScript. We would like to rewrite visitor-as into AssemblyScript to make it more self-hosted which should contribute to the autonomy of the AssemblyScript ecosystem, and help with the quality of visitor-as.

### Awards

NEAR offers anyone who implements it 1000 NEAR tokens.

**Comment by yjhmelody** [comment]
Hi, I am interested in doing these things, but it will definitely simplify some functions of vistor-as, after all, AS does not have any dynamic characteristics. However, it is relatively easy to achieve the effect of the TS version when it is used, in my opinion.
What is the process of applying for these grants projects? What kind of strict regulations are there?

**Comment by yjhmelody** [comment]
By the way, visitor-as has some design issues and implementation issues, I personally think.

**Comment by willemneal** [comment]
Hey @yjhmelody! Commenting here is the first step and I agree with the need for an improved design. 

Here is a PR that I started with attempting to compile a transformer to Wasm: https://github.com/willemneal/visitor-as/pull/12, which you already commented on.

That PR is stale now and copying over the source from `assemblyscript`'s main repo isn't optimal since the source becomes out of date when the AST is updated.

So the next step is to open a new PR and comment the link for it here.

If you aren't on discord yet you should join at [here](https://near.chat).

**Comment by yjhmelody** [comment]
Maybe a new repo is needed to implement the AS version of the visitor library

**Comment by yjhmelody** [comment]
I'm on discord already, but where should this be discussed.

**Comment by yjhmelody** [comment]
I thought about it carefully, maybe it is necessary to temporarily end this grant task. I am still not very clear about the necessity of supporting AS to write transforms, because both in terms of syntax and ecology, AS is not as convenient as TS to write transforms.
And because the middle and back ends of the AS compiler rely heavily on the nodejs library, the program/module in the transform interface cannot become an AS object. It is also unrealistic to pass all these high-level interfaces to AS through WASM, and the maintainability and performance of the code may become worse.

It seems that AS official staff did not respond to the issue.

**Comment by yjhmelody** [comment]
Maybe we can focus on improving as-bignum, this library is very important.

**Comment by willemneal** [comment]
@yjhmelody I agree that trying to make AS transforms without the aid of the core team is not feasible for this grant.  However, it is still crucial for transforms to be explicit in the source code. I have an idea of how to do this.

In your entry file (or any entry file) you do `import "libraryThatNeedsTransform"`.  This could be an empty file or the file that defines globals as we have it now.  We then have a single transform that is the transform loader. This is a pretty unique import style so, instead of checking every import on entry files you just need to check these kind (or just check all imports).  You then track down the package.json of the package, which then has a transform: path/to/file.js.  Or some other way.

Any thoughts?  Does this make sense?

**Comment by yjhmelody** [comment]
@willemneal 
Do you mean a transform written only for a certain AS file (or AS lib)? When importing an AS file, the version modified by this transform is actually imported, right?
While it may seem to simplify some of the specific purpose of writing transform, but feel there is no obvious benefit. I want to know what you will achieve with this feature? Is it for conditional compilation or just to generate many complex APIs for certain libraries?

**Comment by willemneal** [comment]
Let me explain in more details.

The problem that needs to be solved is how to add transforms without using CLI/config.  Rather the source code should be explicit that a transform is needed.  My idea is to create one transform that acts as the transform loader.

e.g.

###  index.ts
```ts
import "lib-with-transform"
```

### lib-with-transform/assembly/index.ts
```ts
// Could be empty or define global variables

```

`lib-with-transform` could have the following project structure:

assembly/index.ts
src/index.ts <---- transform source
index.js <--- transform artifact

`package.json`
```json
"transform": "index.js",
```

The loader transform can then look up in the `package.json` the transform that needs to be loaded.  The other useful aspect of the transform loader is that it can ensure that only one instance of a transform is used if multiple dependencies require the same one.

Of course there are other ways this could be achieved and it is just an idea of how to do it.  Having transforms in AS, while a good goal, is really just one solution to how to load a transform that is made explicit in the source.

So to recap, for all imports to non-project files the transform loader would look up whether there is an associated transformer. This transformer would act the same as if it were added via cli `--transform ...` or through `asconfig.json`.  So now you'll only need `--transform "transform-loader"`.

Once this is done we can look to convincing the core team that this should be added to the compiler and then there won't be a need for adding the transform loader at all.
